usr/opt/OSCX200/src/man/XI/XChProp.man
.\"
.\" *****************************************************************
.\" *                                                               *
.\" *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
.\" *                                                               *
.\" *   All Rights Reserved.  Unpublished rights  reserved  under   *
.\" *   the copyright laws of the United States.                    *
.\" *                                                               *
.\" *   The software contained on this media  is  proprietary  to   *
.\" *   and  embodies  the  confidential  technology  of  Digital   *
.\" *   Equipment Corporation.  Possession, use,  duplication  or   *
.\" *   dissemination of the software and media is authorized only  *
.\" *   pursuant to a valid written license from Digital Equipment  *
.\" *   Corporation.                                                *
.\" *                                                               *
.\" *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
.\" *   by the U.S. Government is subject to restrictions  as  set  *
.\" *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
.\" *   or  in  FAR 52.227-19, as applicable.                       *
.\" *                                                               *
.\" *****************************************************************
.\"
.\"
.\" HISTORY
.\"
.ds xL Programming With Xlib
.TH XChangeDeviceDontPropagateList 3X11 "Release 5" "X Version 11" "X FUNCTIONS"
.SH NAME
XChangeDeviceDontPropagateList, XGetDeviceDontPropagateList \- query or change the dont-propagate-list for extension devices
.SH SYNTAX
XChangeDeviceDontPropagateList\^(\^\fIdisplay\fP, \fIwindow\fP\^,
\fIcount\fP\^, \fIevent_list\fP\^, \fImode\fP\^)
.br
      Display *\fIdisplay\fP\^;
.br
      Window *\fIwindow\fP\^; 
.br
      int *\fIcount\fP\^; 
.br
      XEventClass *\fIevent_list\fP\^; 
.br
      int \fImode\fP\^; 
.br
XEventClass *XGetDeviceDontPropagateList\^(\^\fIdisplay\fP, \fIwindow\fP\^, 
\fIcount\fP\^)
.br
      Display *\fIdisplay\fP\^;
.br
      Window *\fIwindow\fP\^; 
.br
      int *\fIcount\fP\^; 
.SH ARGUMENTS
.TP 12
.I display
Specifies the connection to the X server.
.TP 12
.I window
Specifies the window whose dont-propagate-list is to be queried or modified.
.TP 12
.I event_list
Specifies a pointer to a list of event classes.
.TP 12
.I mode
Specifies the mode.
You can pass 
\fIAddToList\fP , 
or
\fIDeleteFromList\fP.
.TP 12
.I count
Specifies the number of event classes in the list.
.SH DESCRIPTION
The \fIXChangeDeviceDontPropagateList\fP request modifies the list
of events that should not be propagated to ancestors of the event window.
This request allows extension events to be added to or deleted from
that list.  By default, all events are propagated to ancestor windows.
Once modified, the list remains modified for the life of the window.
Events are not removed from the list because the client that added them
has terminated.
.P
Suppression of event propagation is not allowed for all input extension
events.  If a specified event class is one that cannot be suppressed,
a \fIBadClass\fP error will result.  Events that can be suppressed
include \fIDeviceKeyPress\fP, \fIDeviceKeyRelease\fP, \fIDeviceButtonPress\fP,
\fIDeviceButtonRelease\fP, \fIDeviceMotionNotify\fP, \fIProximityIn\fP, 
and \fIProximityOut\fP.
.P
\fIXChangeDeviceDontPropagateList\fP
can generate a \fIBadDevice\fP, \fIBadClass\fP, or \fIBadValue\fP error.
.P
The \fIXGetDeviceDontPropagateList\fP request queries the list
of events that should not be propagated to ancestors of the event window.
.P
\fIXGetDeviceDontPropagateList\fP
can generate a \fIBadClass\fP  or \fIBadWindow\fP error.
.SH DIAGNOSTICS
.TP 12
\fIBadDevice\fP
An invalid device was specified.  The specified device does not exist or has 
not been opened by this client via \fIXOpenInputDevice\fP.  This error may
also occur if some other client has caused the specified device to become
the X keyboard or X pointer device via the \fIXChangeKeyboardDevice\fP or
\fIXChangePointerDevice\fP requests.
.TP 12
\fIBadValue\fP
Some numeric value falls outside the range of values accepted by the request.
Unless a specific range is specified for an argument, the full range defined
by the argument's type is accepted.  Any argument defined as a set of
alternatives can generate this error.
.TP 12
\fIBadWindow\fP
An invalid window id was specified.
.TP 12
\fIBadClass\fP
An invalid event class was specified.
.SH "SEE ALSO"
.br
\fI\*(xL\fP


usr/opt/OSCX200/src/lib/X/XChProp.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
/* $XConsortium: XChProp.c,v 11.22 91/01/06 11:44:22 rws Exp $ */
/* Copyright    Massachusetts Institute of Technology    1986	*/

/*
Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation, and that the name of M.I.T. not be used in advertising or
publicity pertaining to distribution of the software without specific,
written prior permission.  M.I.T. makes no representations about the
suitability of this software for any purpose.  It is provided "as is"
without express or implied warranty.
*/

#include "Xlibint.h"

#if NeedFunctionPrototypes
XChangeProperty (
    register Display *dpy,
    Window w,
    Atom property,
    Atom type,
    int format,  /* 8, 16, or 32 */
    int mode,  /* PropModeReplace, PropModePrepend, PropModeAppend */
    _Xconst unsigned char *data,
    int nelements)
#else
XChangeProperty (dpy, w, property, type, format, mode, data, nelements)
    register Display *dpy;
    Window w;
    Atom property, type;
    int format;  /* 8, 16, or 32 */
    int mode;  /* PropModeReplace, PropModePrepend, PropModeAppend */
    unsigned char *data;
    int nelements;
#endif
    {
    register xChangePropertyReq *req;
    register long len;

    LockDisplay(dpy);
    GetReq (ChangeProperty, req);
    req->window = w;
    req->property = property;
    req->type = type;
    req->format = format;
    req->mode = mode;
    if (nelements < 0) {
	req->nUnits = 0;
	req->format = 0; /* ask for garbage, get garbage */
    } else
	req->nUnits = nelements;
    
    switch (format) {
      case 8:
	len = req->length + (((long)nelements + 3)>>2);
	if (len <= 65535) {
	    req->length = len;
	    Data (dpy, (char *)data, nelements);
	} /* else force BadLength */
        break;
 
      case 16:
	len = req->length + (((long)nelements + 1)>>1);
	if (len <= 65535) {
	    req->length = len;
	    len = (long)nelements << 1;
	    Data16 (dpy, (short *) data, len);
	} /* else force BadLength */
	break;

      case 32:
	len = req->length + (long)nelements;
	if (len <= 65535) {
	    req->length = len;
	    len = (long)nelements << 2;
#if ( defined (__osf__) && defined(__alpha) )
	    {
	    /* sizeof(CARD32) != sizeof(long) */
	    int i;
	    CARD32 *cptr;
	    unsigned long *lptr = (unsigned long *) data;
	    /* per clarification of spec.. unsigned longs are passed
	       into this function but only the lower 32 bits will
	       actually be used - ddhill */
	    cptr = (CARD32 *)_XAllocScratch(dpy,(long)sizeof(CARD32) * len);
	    for(i=0;i<nelements;i++)
		cptr[i] = lptr[i];
	    Data32 (dpy, (long *) cptr, len);
	    }
	    
#else
	    Data32 (dpy, (long *) data, len);
#endif
	} /* else force BadLength */
	break;

      default:
        /* BadValue will be generated */ ;
      }

    UnlockDisplay(dpy);
    SyncHandle();
    }





